src/i2c/slave.rs

```diff
@@ -409,26 +510,141 @@ impl I2cSlave<'_, Blocking> {
 
         let stat = i2c.stat().read();
         if stat.slvdesel().is_deselected() {
-            // clear the deselect bit
-            i2c.stat().write(|w| w.slvdesel().deselected());
-            return Ok(Response::Complete(xfer_count));
+            // inhibit drop
+            core::mem::forget(self);
+            return Ok(ReadResult::Complete(xfer_count));
         } else if stat.slvstate().is_slave_address() {
             // Handle restart after read
-            return Ok(Response::Complete(xfer_count));
+            // inhibit drop
+            core::mem::forget(self);
+            return Ok(ReadResult::Complete(xfer_count));
         } else if stat.slvstate().is_slave_transmit() {
             // Master is still expecting data, transaction incomplete
-            return Ok(Response::Pending(xfer_count));
+            return Ok(ReadResult::Partial(self));
         }
 
         // We should not get here
         Err(TransferError::WriteFail.into())
     }
+
+    /// Finish the entire read transaction, providing the overrun character once the buffer runs out
+    pub fn handle_complete(self, buffer: &[u8], ovc: u8) -> Result<usize> {
+        match self.handle_part(buffer)? {
+            ReadResult::Partial(mut this) => {
+                let mut total = buffer.len();
+                loop {
+                    match this.handle_part(&[ovc])? {
+                        ReadResult::Partial(handler) => {
+                            this = handler;
+                            total += 1;
+                        }
+                        ReadResult::Complete(extra) => {
+                            return Ok(total + extra);
+                        }
+                    }
+                }
+            }
+            ReadResult::Complete(size) => Ok(size),
+        }
+    }
+}
+
+impl Drop for BlockingI2cSlaveRead<'_> {
+    fn drop(&mut self) {
+        let i2c = self.info.regs;
+        if self.should_ack_addr {
+            i2c.slvctl().modify(|_, w| w.slvnack().nack());
+        } else {
+            // complete the read with all-ones overrun
+            loop {
+                // Block until something happens
+                blocking_poll(self.info);
```

Comment:
I think we should move handling the fill character into the interrupt handler, blocking in `drop` seems a bin unexpected to me.

Reply:
For async this is indeed the better approach, however the blocking interface doesn't have an interrupt handler.

Reply:
Fixed this in a different way, the drops are now short even for the blocking implementation.